### PR TITLE
FEAT-#4827: Implement `infer_types` dataframe algebra operator

### LINKED ISSUE #4827: FEAT: Implement `infer_types` dataframe algebra operator
The Pandas [`infer_objects`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.infer_objects.html) function (as well a the [`infer_objects`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.convert_dtypes.html) option of `convert_dtypes`) can both be made more efficient by implementing the `infer_types` dataframe algebra operator.

### PR DESCRIPTION
**Pull request by noloerino**
## What do these changes do?

This implements the `infer_types(col_labels)` DataFrame algebra operator, which attempts to infer more specific datatypes for the columns specified by `col_labels`. This is used to implement the pandas `df.infer_objects()` method, which infers datatypes for columns that have an `object` dtype.

The `infer_types` operator can also be used for `df.convert_dtypes`, but this PR avoids that for now since the Pandas API has some extra flags that are hard to mimic as is.

Initial perf measurements (8-core M1 macOS, 16GB RAM):
```
data = np.random.randint(0, 100, size=(2 ** 16, 2 ** 12))
df = pd.DataFrame(data, dtype=object)
with timer():
    result = df.infer_objects().dtypes
```
**Pandas:** 67.674s
**Ray, current PR (46dcf):** 15.315s
**Dask, current PR:** 13.779s
**Ray, current master (b0465):** 65.844s
**Dask, current master:** 65.129s

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #4827 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date 
- [x] added (Issue Number: PR title (PR Number)) and github username to release notes for next major release

### CODE REVIEW

**Comment by vnlitvinov** [comment]
@noloerino not having looked at code yet, but why Dask isn't getting so blazing fast compared to Ray?

**Comment by noloerino** [comment]
I'm not sure, I'll have to look into it. The time seems to indicate it's somehow defaulting to pandas, but I'm unsure what code path would cause that.

**Comment by noloerino** [comment]
Turns out I mistakenly left my benchmarking code with vanilla pandas when I tested Dask -- I edited the correct run into the post, and Dask is just as fast as Ray here.

Should I label this PR as PERF instead of FEAT?

**Comment by mvashishtha** [comment]
> Should I label this PR as PERF instead of FEAT?

I think "FEAT" is correct here. "PERF" should apply to PRs that are meant to improve an existing Modin feature.

File: docs/supported_apis/dataframe_supported.rst [hunk 1]
```diff
@@ -185,7 +185,7 @@ default to pandas.
 | ``iloc``                   | `iloc`_                   | Y                      | **Omnisci**: ``P``, read access fully supported,   |
 |                            |                           |                        | write access: no row and 2D assignments support    |
 +----------------------------+---------------------------+------------------------+----------------------------------------------------+
-| ``infer_objects``          | `infer_objects`_          | D                      |                                                    |
+| ``infer_objects``          | `infer_objects`_          | Y                      |                                                    |
```
Comment (vnlitvinov) [inline docs/supported_apis/dataframe_supported.rst:188]: that change is questionable, I think it should be `P` for `Partial`, unless you say that OmniSci can do this, too?

File: docs/supported_apis/series_supported.rst [hunk 2]
```diff
@@ -215,7 +215,7 @@ the related section on :doc:`Defaulting to pandas </supported_apis/index>`.
 +-----------------------------+---------------------------------+----------------------------------------------------+
 | ``index``                   | Y                               |                                                    |
 +-----------------------------+---------------------------------+----------------------------------------------------+
-| ``infer_objects``           | D                               |                                                    |
+| ``infer_objects``           | Y                               |                                                    |
```
Comment (vnlitvinov) [inline docs/supported_apis/series_supported.rst:218]: same for OmniSci

File: modin/core/dataframe/pandas/dataframe/dataframe.py [hunk 3]
```diff
@@ -1648,21 +1649,51 @@ def fold(self, axis, func):
...
         Returns
         -------
         PandasDataframe
             A new PandasDataframe with the inferred schema.
         """
-        pass
+        # Compute dtypes on the specified columns, and then set those dtypes on a new frame
+        new_cols = self.take_2d_labels_or_positional(col_labels=col_labels)
+        new_cols_dtypes = new_cols.tree_reduce(0, pandas.DataFrame.infer_objects).dtypes
+        new_dtypes = self.dtypes.copy()
```
Comment (pyrito) [inline modin/core/dataframe/pandas/dataframe/dataframe.py:1687]: do we have to do a copy here or can we just update the dtypes in place with the same reference?

Comment (noloerino) [reply modin/core/dataframe/pandas/dataframe/dataframe.py:1687]: Yes, or else `infer_types` would mutate the dtypes of the original dataframe.

```
>>> df = pd.DataFrame({"a": ["s", 2, 3]}).iloc[1:]
>>> df.dtypes
a    object
dtype: object
>>> df.infer_objects()
<modin.core.storage_formats.pandas.query_compiler.PandasQueryCompiler object at 0x7fb0124b7880>
>>> df.dtypes
a    int64
dtype: object
```

File: docs/supported_apis/dataframe_supported.rst (hunk 1 above)
Comment (noloerino) [reply docs/supported_apis/dataframe_supported.rst]: I've added a note indicating ``D`` for Omnisci.

File: modin/pandas/test/dataframe/test_map_metadata.py [hunk 4]
```diff
@@ -571,6 +571,15 @@ def test_astype_category_large():
     assert modin_result.dtypes.equals(pandas_result.dtypes)
 
 
+def test_infer_objects_single_partition():
+    data = {"a": ["s", 2, 3]}
+    modin_df = pd.DataFrame(data).iloc[1:]
+    pandas_df = pandas.DataFrame(data).iloc[1:]
+    modin_result = modin_df.infer_objects()
+    pandas_result = pandas_df.infer_objects()
+    assert modin_result.dtypes.equals(pandas_result.dtypes)
```
Comment (pyrito) [inline modin/pandas/test/dataframe/test_map_metadata.py:580]: I was going to suggest that we use `df_equals` to check this but then realized that we do not check `dtypes` there I think...oops...

**Review by pyrito — APPROVED** [review]
LGTM!

File: modin/pandas/test/dataframe/test_map_metadata.py (hunk 4 above)
Comment (vnlitvinov) [reply modin/pandas/test/dataframe/test_map_metadata.py:580]: this should eventually become a feature: https://github.com/modin-project/modin/issues/3804

**Review by vnlitvinov — APPROVED** [review]
LGTM, thanks @noloerino!

File: modin/pandas/test/dataframe/test_map_metadata.py (hunk 4 above)
Comment (pyrito) [reply modin/pandas/test/dataframe/test_map_metadata.py:580]: Yes, we should fix that soon, I think we might find a bunch of bugs.

**Pull request merged**